### PR TITLE
Doe jacobian remake

### PR DESCRIPTION
**Pull request by Osburg**
This PR adds:
- Computation of objective function gradient using autodiff and formulaic's symbolic differentiation of model matrices. Now supports arbitrary model terms and not only polynomial model terms of order <= 3. Significant speedup of gradient evaluation in case of large models with a large number of experiments.
- Restructuring the objective function and its gradient. A new class `DOptimality` now contains both constraint evaluation and constraint jacobian evaluation.
- `__call__()` and `jacobian()` methods for `NChooseKConstraint` using narrow Gaussians as discussed in #145 . 
- Option in doe to pass NChooseK constraints to IPOPT (using the new constraint evaluation and jacobian) instead fixing a corresponding number of variables in each experiment to zero **before** the optimization. Possibly to be removed in the future since the new option does not seem to work nicely so far.

### CODE REVIEW

File: bofire/strategies/doe/utils.py [hunk 1]
```diff
@@ -519,12 +531,13 @@ def nchoosek_constraints_as_bounds(
     # bounds without NChooseK constraints
     bounds = np.array(
         [
-            (p.lower_bound, p.upper_bound)
+            p.bounds
             for p in domain.inputs
             if not (
                 isinstance(p, CategoricalInput)
```
Comment (jduerholt) [inline bofire/strategies/doe/utils.py:537]: same here, use the getters.

**Comment by jduerholt** [comment]
Out of curiosity: What do you mean with the new nchoosek option does not seem to work nicely? What is happening if you use it?

File: bofire/strategies/doe/objective.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        J = np.sum(J, axis=-1)
+
+        return J.flatten()
+
+    def _convert_input_to_model_tensor(self, x: np.ndarray) -> Tensor:
+        X = pd.DataFrame(
+            x.reshape(len(x.flatten()) // self.n_vars, self.n_vars), columns=self.vars
+        )
+        X = self.model.get_model_matrix(X)
+        return torch.tensor(
```
Comment (jduerholt) [inline bofire/strategies/doe/objective.py:145]: Use here the tkwargs arguments from `bofire.utils.torch_tools`

**Comment by Osburg** [comment]
About the nchoosek constraints not working properly. I explained my findings [here](https://github.com/experimental-design/bofire/issues/145#issuecomment-1501089778).

File: bofire/strategies/doe/objective.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    def __init__(
+        self,
+        domain: Domain,
+        model: Formula,
+        n_experiments: int,
+        delta: float = 1e-7,
+    ) -> None:
+        """
+        Args:
+            domain (Domain): An opti domain defining the DoE domain together with model_type.
```
Comment (jduerholt) [inline bofire/strategies/doe/objective.py:57]:
```suggestion
            domain (Domain): An domain defining the DoE domain together with model_type.
```

File: bofire/strategies/doe/design.py [hunk 4]
```diff
@@ -151,27 +116,37 @@ def find_local_max_ipopt(
...
+        bounds = nchoosek_constraints_as_bounds(domain, n_experiments)
+    else:
+        bounds = np.array(
+            [
+                p.bounds
+                for p in domain.inputs
+                if not (
+                    isinstance(p, CategoricalInput)
+                    or isinstance(p, ContinuousOutput)
+                    or isinstance(p, MolecularInput)
```
[the remark this answers is not in the document]
Comment (Osburg) [reply bofire/strategies/doe/design.py]: Thanks for the hint. Using the getter now.

**Review by jduerholt — APPROVED** [review]
Looks good to me. Thanks!

**Comment by jduerholt** [comment]
@dlinzner-bcs @DavidWalz : Do you also want to have a look, or should I just merge?

**Comment by jduerholt** [comment]
I think, we should just merge it ;)
